### DISCUSSION TITLE
Missing avalonia.js in Browser WASM Application

### DISCUSSION THREAD

**Discussion by Tryanks** [AvaloniaUI/Avalonia#15665] [Q&A] [answered]
My Avalonia version is `11.1.0-beta2`, using `.NET 8`.
I attempted to launch my browser application, but it got stuck on the loading screen. So I created a new project directly using a template, as shown:
![83d02d1944b5f5c66bcfd7ad257c4c61](https://github.com/AvaloniaUI/Avalonia/assets/49746336/f5d23f7a-0b0c-471c-aefc-f2fcdb9bbf3c)
_Note that after creating the template, I upgraded the dependencies to beta2._

When I directly started the newly created template project Test.Browser, I still encountered the same result.
Upon opening F12, the console displayed the following error:
![image](https://github.com/AvaloniaUI/Avalonia/assets/49746336/85b6e86c-faed-40f3-b88d-63e48a4f217f)
When I checked the Test/Test.Browser/bin/Release/net8.0-browser/wwwroot/_framework path, I indeed could not find avalonia.js.

**Comment by Tryanks** [comment, marked as answer]
That's cause my `Wasm-Tools` version is `wasm-tools-net7`.
Run
```
dotnet workload install wasm-tools
```
Update to version 8. (Matched to .NET 8.0)

And then avalonia.js will generate correctly.

**Reply by tbayart** [reply]
not working for me
cleaned solution, rebuild, reboot, still the same 404 because of missing avalonia.js

**Reply by maxkatz6** [reply]
@tbayart https://docs.avaloniaui.net/docs/guides/platforms/how-to-use-web-assembly

**Comment by AlexeyTliss** [comment]
For people who still are having this issue:
Please note that command `dotnet workload install wasm-tools` installs wasm-tools for your default `.NET` version, which, in fact, can be different from your project version. For example we have project targeting Net8, but MS made Net9 default in latest visual studio, so Avalonia.Browser won't start for us with error "missing avalonia.js"

To fix this, install specific version of wasm-tools that matches your project version, which are:
`dotnet workload install wasm-tools-net7`
`dotnet workload install wasm-tools-net8`
`dotnet workload install wasm-tools-net9`
`dotnet workload install wasm-tools-net10`
Cheers!

**Reply by sds-eng** [reply]
I am having the same issue on the standard Avalonia web browser template generated in VS 2022. 

I get `Workload installation failed: Workload ID wasm-tools-net9 is not recognized.` when trying to run `dotnet workload install wasm-tools-net9`. Am I missing something obvious? 

It looks like there isn't a `wasm-tools` for anything later than `net8`:

```
dotnet workload search wasm-tools-net

Workload ID          Description
------------------------------------------------------------
wasm-tools-net6      .NET WebAssembly build tools for net6.0
wasm-tools-net7      .NET WebAssembly build tools for net7.0
wasm-tools-net8      .NET WebAssembly build tools for net8.0
```

**Reply by tbayart** [reply]
try this one :
`dotnet workload install wasm-tools`

**Reply by tbayart** [reply]
```
dotnet workload list

Installed Workload Id      Manifest Version       Installation Source          
-------------------------------------------------------------------------------
android                    35.0.92/9.0.100        SDK 9.0.300, VS 17.14.36408.4
ios                        18.5.9215/9.0.100      SDK 9.0.300, VS 17.14.36408.4
maccatalyst                18.5.9215/9.0.100      SDK 9.0.300, VS 17.14.36408.4
maui-windows               9.0.82/9.0.100         SDK 9.0.300, VS 17.14.36408.4
wasm-tools                 9.0.8/9.0.100          SDK 9.0.300                  
wasm-tools-net8            9.0.8/9.0.100          SDK 9.0.300                  
```

**Reply by Tryanks** [reply]
> I get `Workload installation failed: Workload ID wasm-tools-net9 is not recognized.` when trying to run `dotnet workload install wasm-tools-net9`. Am I missing something obvious?
> 
> It looks like there isn't a `wasm-tools` for anything later than `net8`

I believe that only after .NET 10 is released will wasm-tools-net9 appear in the workload list, representing the workload specific to the older version. Meanwhile, wasm-tools itself always corresponds to the latest version currently under maintenance. By this logic, wasm-tools-net10 will only appear when .NET 11 is released.

**Reply by sds-eng** [reply]
Thanks @Tryanks `dotnet workload install wasm-tools` did the trick and that makes sense why we wouldn't see a `net9` version when listing.

My underlying problem however, seemed to stem from using an old Avalonia template. My Visual Studio shows two "cross platform" options when creating a new project. I selected the "other" one and the project the template created ran without throwing the "failed to load dynamically imported module". 

Thanks for the feedback though, this did help me get unstuck!
